plugin/trino-hudi/src/main/java/io/trino/plugin/hudi/query/HudiReadOptimizedFileLister.java

```diff
@@ -0,0 +1,227 @@
+                allPartitionInfoList = relativePartitionPathList.stream()
+                        .map(relativePartitionPath ->
+                                buildHudiPartitionInfo(
+                                        useMetastoreForPartitions,
+                                        Option.of(relativePartitionPath),
+                                        Option.empty(),
+                                        Option.of(partitionValueExtractor),
+                                        partitionColumns,
+                                        partitionColumnHandles,
+                                        tableHandle.getPartitionPredicates(),
+                                        hiveTable,
+                                        hiveMetastore))
+                        .collect(Collectors.toList());
+            }
+            catch (HoodieIOException e) {
+                log.warn("Cannot infer partition value extractor. Falling back to using Hive Metastore for table %s. %s", tableName, e.getMessage());
+            }
+        }
+
+        if (isNull(allPartitionInfoList)) {
+            // Use Hive partition names and other context to construct
+            // HiveHudiPartitionInfo instances
+            if (isNull(hivePartitionNames)) {
+                hivePartitionNames = partitionColumns.isEmpty()
+                        ? Collections.singletonList("")
+                        : getPartitionNamesFromHiveMetastore(partitionKeysFilter);
+            }
+
+            allPartitionInfoList = hivePartitionNames.stream()
+                    .map(hivePartitionName ->
+                            buildHudiPartitionInfo(useMetastoreForPartitions,
+                                    Option.empty(),
+                                    Option.of(hivePartitionName),
+                                    Option.empty(),
+                                    partitionColumns,
+                                    partitionColumnHandles,
+                                    tableHandle.getPartitionPredicates(),
+                                    hiveTable,
+                                    hiveMetastore))
+                    .collect(Collectors.toList());
+        }
+
+        List<HudiPartitionInfo> filteredPartitionInfoList = allPartitionInfoList.stream()
+                .filter(partitionInfo -> partitionInfo.getHivePartitionKeys().isEmpty() || partitionInfo.doesMatchPredicates())
+                .collect(Collectors.toList());
+
+        log.debug("Get partitions to scan in %d ms (useMetastoreForPartitions: %s): %s", timer.endTimer(), useMetastoreForPartitions, filteredPartitionInfoList);
+
+        return filteredPartitionInfoList;
+    }
+
+    @Override
+    public List<FileStatus> listStatus(HudiPartitionInfo partitionInfo)
+    {
+        return fileSystemView.getLatestBaseFiles(partitionInfo.getRelativePartitionPath())
+                .map(baseFile -> {
```

Comment:
nit: Extract the file status conversion and exception handling to a private method to improve readability of the method.